design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,173 @@
+# Uncommitted Garbage Collector
+
+## Motivation
+
+Uncommitted data which is no longer referenced (due to any staged data deletion or override, reset branch etc.) is not being deleted by lakeFS.
+This may result in excessive storage usage and possible compliance issues.
+To solve this problem two approaches were suggested:
+1. A batch operation performed as part of an external process (GC)
+2. An online solution inside lakeFS
+
+Several attempts for an online solution have been made, most of which are documented [here](https://github.com/treeverse/lakeFS/blob/master/design/rejected/hard-delete.md).
+This document will describe the **offline** GC process for uncommitted objects.
+
+## Design
+
+Garbage collection of uncommitted data will be performed using the same principles of the current GC process.
+The basis for this is a GC Client (i.e. _Spark_ job) consuming objects information from both lakeFS and directly from the underlying object storage,
+and using this information to determine which objects be deleted from the namespace.
+
+The GC process is composed of 3 main parts:
+1. Listing namespace objects
+2. Listing of lakeFS repository committed objects
+3. Listing of lakeFS repository uncommitted objects
```

Comment:
An object that's in case 3 means that there's an active branch that holds the uncommitted object, or is it any uncommitted object under that repository ("dangling" or not)?